### PR TITLE
Test the binding package and fixed some minor issues

### PR DESCRIPTION
**Pull request by gtudan**
I was doing some low-energy work and started testing the converter package. There are some edge-cases that we don't handle correctly:

* Float and Double usually return NaN - but not always
* There were uncaugt NumberFormatExceptions with BigDecimal and BigInteger

There are still some flaky tests, which I left commented out - the parse behavior slightly changed from Java 8 to 11. It's not that bad, but a little supprising ("NaN" is a valid Number and won't trigger binding errors).

It's a little strange that nearly all converters handle numbers, but we don't have a converter that supports the abstract `Number`-Type as a result. Should we pick one that handles this (Double?)

We also forgot about `Short` should we add a converter for it?

### CODE REVIEW

**Comment by gtudan** [comment]
Sorry for the commit noise - I had to redo some commits since I was using the wrong email. :-( 

I added a converter for short like Daniel suggested. That should be all of the most basic number types (maybe with the exception of byte, but that seemed a bit to exotic as numeric).

**Comment by chkal** [comment]
Sorry for the delay. I hope to find some time later today to review the PR.

File: core/src/main/java/org/eclipse/krazo/binding/convert/impl/DoubleConverter.java [hunk 1]
```diff
@@ -37,7 +37,7 @@ public boolean supports(Class<Double> rawType) {
     @Override
     public ConverterResult<Double> convert(String value, Class<Double> rawType, Locale locale) {
 
-        Double defaultValue = Double.TYPE.equals(rawType) ? 0.0 : null;
+        Double defaultValue = Double.TYPE.equals(rawType) ? 0.0 : Double.NaN;
```
Comment (chkal) [inline core/src/main/java/org/eclipse/krazo/binding/convert/impl/DoubleConverter.java]: I'm not sure about this one. Actually I think this violates the spec. The spec states in "3.4.1 Numeric types":

> Empty strings are either converted to null or to the default value of the corresponding primitive data type. 

The problematic part about our data binding integration is that JAX-RS ALWAYS expects a result from a converter if it doesn't throw an exception (and therefore aborts the request lifecycle with a 400 status codes). Therefore, our converter MUST return a result which JAX-RS can bind to the corresponding field/parameter even in case of validation errors and empty submissions.

Comment (gtudan) [reply core/src/main/java/org/eclipse/krazo/binding/convert/impl/DoubleConverter.java]: If that's what the spec says fine :-) I'll revert this back to null

I was aiming for a more consistent behavior on the edge cases. Right now we return the NaN in most of those cases (with float and double).

File: core/src/main/java/org/eclipse/krazo/binding/convert/impl/BigDecimalConverter.java [hunk 2]
```diff
@@ -46,7 +46,7 @@ public boolean supports(Class<BigDecimal> rawType) {
                             .orElse(null)
             );
 
-        } catch (ParseException e) {
+        } catch (ParseException | NumberFormatException e) {
```
[the remark this answers is not in the document]
Comment (gtudan) [reply core/src/main/java/org/eclipse/krazo/binding/convert/impl/BigDecimalConverter.java]: NumberFormat does not throw `NumberFormatException`, but the constructors of BigInteger and BigDecimal do. I didn't get the BigDecimalConverter to throw this, but BigIntegerConverter will, if you pass a decimal (like 12.3).

[the remark this answers is not in the document]
Comment (chkal) [reply core/src/main/java/org/eclipse/krazo/binding/convert/impl/BigDecimalConverter.java]: Ah, good catch. Thanks a lot!

I wonder if we should improve `BigIntegerConverter` to handle the case of decimals better!? We could first create a `BigDecimal`, then round it to remove all fraction digits and then convert it to an `BitInteger`. If the user enters `12.9` and the value is bound to a `BigInteger`, binding a value of `13` (rounded via `RoundingMode.HALF_UP`) sounds reasonable to me. Thoughts?

File: core/src/main/java/org/eclipse/krazo/binding/convert/impl/DoubleConverter.java (hunk 1 above)
Comment (chkal) [reply core/src/main/java/org/eclipse/krazo/binding/convert/impl/DoubleConverter.java]: Really? See: 

https://github.com/eclipse-ee4j/krazo/blob/05d4c03bf8f7e915478022727dedc3b3471365d8/core/src/main/java/org/eclipse/krazo/binding/convert/impl/DoubleConverter.java#L40

The basic idea was to return the data type "initialization default value" in case of empty strings or validation errors. As described above, JAX-RS requires us to return a value or it will abort the request. Actually the user would expect that no binding happens at all if a binding error occurs, but JAX-RS forces us to return a value from the converter. And by returning the data type default value, it almost looks like JAX-RS doesn't perform any data binding in these cases, also it actually does (but the user won't notice).

I know that this feels dirty, but I didn't find any other way to implement the binding/validation error handling as required by the spec.

**Comment by chkal** [comment]
BTW: It looks like the `ip-validation` check is hanging (again). I already filed [an issue](https://bugs.eclipse.org/bugs/show_bug.cgi?id=544974) regarding this. Let's hope this gets resolved soon, because it currently blocks the merge.

**Comment by chkal** [comment]
According to [this comment](https://bugs.eclipse.org/bugs/show_bug.cgi?id=544974#c1) adding another commit to the PR could fix the problem regarding the hanging ip-validation check. AFAIK rebasing (and therefore updating all the commit hashes) should also fix the issue and may even work better.

File: core/src/main/java/org/eclipse/krazo/binding/convert/impl/BigDecimalConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gtudan) [reply core/src/main/java/org/eclipse/krazo/binding/convert/impl/BigDecimalConverter.java]: I support the idea of converting to BigInteger with loss of precision, but I'm not so sure about the rounding mode. This would mean that BigInteger behaves quiet different than Integer, which just drops the decimal, as long as it's in the precision:

```
jshell> (int) 2.5
$3 ==> 2

jshell> (int) 2.6
$4 ==> 2

jshell> (int) 2.9
$5 ==> 2

jshell> (int) 2.9999999999999999999999999
$6 ==> 3
```

**Comment by gtudan** [comment]
I'd rather leave the BigInteger-Fix to another issue, as I'm not sure about the correct rounding mode. 
That ip-check is really tricky. I tried changing the last commit: didn't help. Lets see what it does if I rebase all the commits.

File: core/src/main/java/org/eclipse/krazo/binding/convert/impl/BigDecimalConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chkal) [reply core/src/main/java/org/eclipse/krazo/binding/convert/impl/BigDecimalConverter.java]: I would prefer using `RoundingMode.HALF_UP`. WDYT? 

Something like (pseudocode, no handling for empty optional yet):

```java
BigDecimal decimal = new BigDecimal( parseNumber(value, locale).toString() );
BigInteger result = new BigInteger( decimal.setScale(0, RoundingMode.HALF_UP).toString() );
```

**Comment by chkal** [comment]
> I'd rather leave the BigInteger-Fix to another issue, as I'm not sure about the correct rounding mode.

Ok, that's fine. Let's get this one in first.

> That ip-check is really tricky. I tried changing the last commit: didn't help. Lets see what it does if I rebase all the commits.

I think the problem is that this PR got created while the EF is migrating from `ip-validation` to `eclipsefdn/eca`. One of the webmasters told me, that he can merge the PR even if `ip-validation` is failing as he has admin permissions for the repository.

**Comment by gtudan** [comment]
> I think the problem is that this PR got created while the EF is migrating from ip-validation to eclipsefdn/eca. One of the webmasters told me, that he can merge the PR even if ip-validation is failing as he has admin permissions for the repository.

Might be. Do you want to ask him, or should I open a new MR?

**Comment by chkal** [comment]
I just asked the EF admins to merge the pull request on our behalf:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=544974#c3
